### PR TITLE
Homie mjcumming WIP

### PR DESCRIPTION
**Pull request by psyciknz**
Added to handle_panel_change to adding of zones.
For each zone the alarm and open properties are created and published via homie.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -24,6 +24,9 @@ chump>=1.6.0
 # IP150 >4.0 Support
 requests>=2.20.0
 #
+# Homie-MQTT Iterface
+git+https://github.com/mjcumming/HomieV3.git
```
Comment (yozik04) [inline requirements.txt]: Probably replace with:
`Homie3>=0.3.0`

**Comment by psyciknz** [comment]
Added a node_filter which can choose the properties of each type to pill (eg zone: "open", "alarm" or system: "vdc", "dc", "battery")

File: paradox/lib/ps.py [hunk 2]
```diff
@@ -14,3 +14,6 @@ def sendEvent(event: dict):
 def sendChange(etype: str, label: str, property: str, value: str, initial: bool=False):
     payload = dict(type=etype, label=label, property=property, value=value, initial=initial)
     pub.sendMessage(PREFIX + "changes", change=payload)
+
+def sendInternal(message: str):
```
Comment (yozik04) [inline paradox/lib/ps.py]: I have just realised that there is a bug. I forgot to add PREFIX. I fixed it in dev.
```
def sendMessage(topicName: str, **msgData):
    pub.sendMessage(PREFIX + topicName, **msgData)
```

Then `sendInternal` is not required.

**Comment by yozik04** [comment]
Can you please try to merge dev into your branch.
I'd do that in a separate branch not to trigger a panic mode.

You will have homeassistant example nearby and latest fixes.

File: paradox/lib/ps.py (hunk 2 above)
Comment (psyciknz) [reply paradox/lib/ps.py]: what is PREFIX?  Does this mean the "enumerated_properties" message will not be sent now?

**Comment by psyciknz** [comment]
> Can you please try to merge dev into your branch.
> I'd do that in a separate branch not to trigger a panic mode.
> 
> You will have homeassistant example nearby and latest fixes.

Ok, I'll merge your dev branch to my dev branch (not where I'm adding homie), then merge my homie code on top.

File: paradox/lib/ps.py (hunk 2 above)
Comment (psyciknz) [reply paradox/lib/ps.py]: So looking at the merged changes it looks like
```
if first_loop:
                first_loop = False
                ps.sendInternal('properties_enumerated')
```
is removed, and I'm using this as a trigger for "I've found all my potential nodes, go process them" and once they're processed I start looking at the changes.

**Comment by psyciknz** [comment]
So I'm looking into this now and trying to figure out how all the mqtt interfaces are all working now that they're subclassed...as my existing one won't call the run method, maybe because it's not async?

So looks like I have to start again.

But with the subclassed mqtt interfaces, in run, you always connect to mqtt....for homie this doesn't work as the connection is handled by the library (which posts all the homie statistics etc).

Any ideas why my run method for my interface is not firing?  What am i misssing.  I see the GSM and other interfaces haven't gone async yet.

**Comment by psyciknz** [comment]
I haven’t movie it to my proper branch yet, but it’s a pretty horrible change for me.   So I finally worked out how to get all the labels.   But what is missing at that point is all the properties?

Have they been found at the point of reading the labels?  I suspect not.

File: paradox/lib/ps.py (hunk 2 above)
Comment (jpbarraca) [reply paradox/lib/ps.py]: This was added to support that functionality and was mandatory for homie (2.0) as we must declare all properties before activating the interface.

Comment (yozik04) [reply paradox/lib/ps.py]: they are actually all internal. So this is just duplicate function there.

See how I do it for homeassistant:
HASS Interface subscribes to:
```
        ps.subscribe(self._handle_status_update, "status_update")
        ps.subscribe(self._handle_labels_loaded, "labels_loaded")

# Handlers
def _handle_labels_loaded(self, data): # Fires when ALL labels are loaded. data has all labels in groups (partiotions, zones, pgms...)
    logger.debug(data)
def _handle_status_update(self, status): #  Fires when all status responses were received. status has all parsed status responses merged into one dict. Useful for partition arming state detection. See hass interface.
    logger.debug(status)
```
I dispatch status update when all statuses were laoded:
```
# in paradox.py
ps.sendMessage('status_update', status=merged)
# ...
# and for labels in panel.py:
ps.sendMessage('labels_loaded', data=self.core.data)
```

**Comment by yozik04** [comment]
Do not subclass AbstractMQTT. Subclass AsyncQueueInterface or ThreadQueueInterface.

in in case AsyncQueueInterface:
```
    async def run(self):
       # do your subscribes
        # ps.subscribe(self._handle_status_update, "status_update")
        # ps.subscribe(self._handle_labels_loaded, "labels_loaded")

        await super().run()
```

**Comment by yozik04** [comment]
Your init stays.Let me help you what needs to stay in mqtt_homie_interface.py:

1. subclass `AsyncQueueInterface`
2. `__init__` stays as is
3. `start` content moves to `run`. Do not forget that `super().run()` needs to be in the end of you `run` function. Drop `start`
4. drop `stop`

All the rest stays the same.

**Comment by yozik04** [comment]
I do not think we need extra threads for interfaces. Anyway all mqtt IO happens in a separate thread. Use async.

**Comment by yozik04** [comment]
Your `properties_enumerated` just notifies you.
My `status_update` passes all properties to the function already.

You do not need to go dig variables in core. Code stays decoupled.

**Comment by yozik04** [comment]
So `status_update` is exactly what you need =) You can check if it is first loop or not inside the interface.

**Comment by yozik04** [comment]
@jpbarraca We need somehow check if EVO and MG/SP property names match between different panels. Any ideas how to simplify this?

**Comment by jpbarraca** [comment]
Send a dump of your properties. It should be just a couple of mosquitto_sub |cut | sort | uniq

**Comment by jpbarraca** [comment]
Check tests/spectra_magellan/properties.log which contains all properties exported by an MG panel.

**Comment by psyciknz** [comment]
> So `status_update` is exactly what you need =) You can check if it is first loop or not inside the interface.

I must be missing something (or must have missed a pull) because in the homeassistant interface (which I'm using as my example), I can't see status_update firing.  I get the update_labels  but neither my new homie or the homeassistant classes fire off the _handle_status_update method

**Comment by psyciknz** [comment]
Also with this version I'm seeing a lot of this:
```
2019-09-06 08:10:18,156 - ERROR    - PAI.paradox.lib.async_message_manager - No handler for message 5
Detail: Container:
    fields = Container:
        data = b'R\x00\x80\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'... (truncated, total 36)
        value = Container:
            po = Container:
                command = 5
                status = Container:
                    reserved = False
                    alarm_reporting_pending = False
                    Winload_connected = True
                    NeWare_connected = False
            validation = 128
            address = 3
            data = b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'... (truncated, total 32)
        offset1 = 0
        offset2 = 36
        length = 36
    checksum = b'\xd5' (total 1)
2019-09-06 08:10:19,077 - ERROR    - PAI.paradox.lib.async_message_manager - No handler for message 5
Detail: Container:
    fields = Container:
        data = b'R\x00\x80\x00\x00\x00\x00\x00\x00\x14\x13\t\x06\x08\n\xd0'... (truncated, total 36)
        value = Container:
            po = Container:
                command = 5
                status = Container:
                    reserved = False
                    alarm_reporting_pending = False
                    Winload_connected = True
                    NeWare_connected = False
            validation = 128
            address = 0
            data = b'\x00\x00\x00\x00\x00\x14\x13\t\x06\x08\n\xd0\x99\x96>\x00'... (truncated, total 32)
        offset1 = 0
        offset2 = 36
        length = 36
    checksum = b'[' (total 1)
2019-09-06 08:10:19,165 - ERROR    - PAI.paradox.lib.async_message_manager - No handler for message 5
Detail: Container:
    fields = Container:
        data = b'R\x00\x80\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'... (truncated, total 36)
        value = Container:
            po = Container:
                command = 5
                status = Container:
                    reserved = False
                    alarm_reporting_pending = False
                    Winload_connected = True
                    NeWare_connected = False
            validation = 128
            address = 3
            data = b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'... (truncated, total 32)
        offset1 = 0
        offset2 = 36
        length = 36
    checksum = b'\xd5' (total 1)
2019-09-06 08:10:19,969 - ERROR    - PAI.paradox.paradox - Lost communication with panel
2019-09-06 08:10:19,973 - INFO     - PAI.paradox.paradox - Disconnecting from the Alarm Panel
2019-09-06 08:10:19,983 - INFO     - PAI.paradox.paradox - Clean Session
2019-09-06 08:10:19,984 - INFO     - PAI.paradox.paradox - Cleaning previous session. Closing connection
2019-09-06 08:10:19,994 - INFO     - PAI.paradox.paradox - Disconnected from the Alarm Panel
2019-09-06 08:10:20,007 - ERROR    - PAI.paradox.connections.ip_connection - Connection to panel was lost
```

**Comment by yozik04** [comment]
@jpbarraca Can you please check what addresses should return from EEPROMRead?

I have changed
https://github.com/jpbarraca/pai/blob/master/paradox/hardware/spectra_magellan/panel.py#L137

to this
https://github.com/jpbarraca/pai/blob/dev/paradox/hardware/spectra_magellan/panel.py#L145

**Comment by yozik04** [comment]
Looks like there is a different expected address....

**Comment by yozik04** [comment]
I have pushed one change to **dev**. Can you try now?

**Comment by yozik04** [comment]
EEPROMRead fixed?

**Comment by yozik04** [comment]
Something is missing? Any help required?

**Comment by psyciknz** [comment]
Yeah just trying to use update labels and status update, which are not really connected.   As labels does not have the attributes (eg zone just has name and label and not the open, or alarm properties). So have to get those from status update.


But the link is an index, so I’m having to store all the labels till status update.


Or is there an object or data structure I can pull from  the panel object or paradox or ps that has a cache of everything?

**Comment by yozik04** [comment]
Yes, your interface needs to store the state:

Take a look at _handle_labels_loaded and _handle_status_update in homeassistant.py

**Comment by yozik04** [comment]
Actually we may need to create a single storage for all that.

**Comment by psyciknz** [comment]
Yeah my main issue is that update labels produces:
Zone: { front door main: {index:1, label:front door main, name: front_door_main}

Has the name and label, but doesn’t list the things like open, alarm, trouble. 


Then under update status I get
zone_open: {1:false}
zone_status: {1: alarm: true, exit_delay: true}

So zone to me is a thing, for homie a node.  But the status updates I get combine the node with a property (open, alarm being properties).

The jpbarraca homie branch was much easier for me to deal with.   This label/status format makes a really messy code block (have a look at the update status of my interface)

**Comment by yozik04** [comment]
I am missing what is complicated here.

Initially you get all zone names. You save that. Then you get first status update you create homie things/nodes. Now you have an initial setup.

Then you listen for `ps.subscribe(self._handle_panel_change, "changes")`
and update homie things/nodes.

**Comment by yozik04** [comment]
I will try to find time this week to make it for homeassistant

**Comment by psyciknz** [comment]
> I am missing what is complicated here.
> 
> Initially you get all zone names. You save that. Then you get first status update you create homie things/nodes. Now you have an initial setup.
> 
> Then you listen for `ps.subscribe(self._handle_panel_change, "changes")`
> and update homie things/nodes.

The issue is in the status update, where the zones and properties being mixed.
Eg a zone dict in the labels contains the names of all the zones.
But in update status, the dict is called zone_open.
And for the power settings, in update labels, it's called system: power....but in update status, the dc, vdc and battery are their own singlular entries.

The lack of consistency is what is making it difficult.  So for the zone, I have to split that it's a zone + open then attach it to index 1 label.  Ok
But for dc, I have to figure out that it's not part of any dict, so I need to assign it to power.

I would find it easier if the labels contained the properties.  Or in update status it was consistent:
```
zone: 1: {"open": true}
power: {1: {"dc": 13, "vdc": 12}
```
rather than
```
zone_open{1: true, 2: false}
dc: 12
```

etc.

**Comment by yozik04** [comment]
Got it. I will try to make up something tomorrow.

**Comment by yozik04** [comment]
Now it is like that:
```
zone: {1: {"open": true, "tamper":true, ...}}
```

@jpbarraca or @psyciknz you need to verify it you do not have issues with latest "dev". If you have. Then I need an aggregated status response from you.

**Pull request closed without merging**